### DISCUSSION TITLE
In-memory read-through cache for PhysicalFileSystemCache

### DISCUSSION THREAD

**Discussion by Piedone** [SixLabors/ImageSharp.Web#349] [Q&A]
I'm [investigating](https://github.com/OrchardCMS/OrchardCore/issues/14859) some static file-serving performance issues for an Orchard Core app running on Azure App Services. It seems that storage IO is being throttled for the app by Azure, due to too many storage operations. ImageSharp's `PhysicalFileSystemCache` is a primary consumer of storage IO, so I'm looking into lowering that.

One option would be to use `AzureBlobStorageImageCache` (since the source images are already stored), and another one I thought about is having an in-memory caching (with `IMemoryCache`) implementation wrapping `PhysicalFileSystemCache`. So, files would still be stored on the file system, but reads (when done repeatedly) would come from memory.

This would help because we have plenty of available RAM, but a very narrow IO budget (with about max 500 operations per second available for everything, including running the OS).

Has anybody done this before? I didn't find any sign of this. Or am I overlooking some aspect of why this is a bad idea?

**Comment by Piedone** [comment]
Anybody with some input, perhaps?

**Comment by deanmarcussen** [comment]
@Piedone why don't you put a cdn in front of the web host for your images?

That's been my general go to for improving performance, and has generally been successful.

**Reply by Piedone** [reply]
I'm doing that and it helps but not completely, since a CDN won't necessarily cache your images for as long as the cache header allows. E.g. Cloudflare removes infrequently accessed files from their cache even if the files' expiration is not due for long (I can't find specifics, but this is a declared behavior; I've seen images being re-requested within the day). This then results in high storage transaction count the next time a page with many images is accessed (which is an issue on Azure App Services: https://github.com/OrchardCMS/OrchardCore/issues/14859).

**Reply by deanmarcussen** [reply]
ok, I have been down a similar road before.

Are you running distributed? 

One issue I found (there where many) where the distributed servers had different image metadata because I'd left all the metadata on the file system.

This caused the CDN to frequently re-request files.

I tried an awful lot of things to get it working consistently, and consistently reading from the CDN cache.

From memory the eventual solution was to host the image metadata files on Azure, but not the actual images themselves.
so using `AzureBlobStorageCacheResolver`

This may not be sufficent for your issues, and you may might be right in considering using the entire set of `AzureBlobStorageImageCache` features.

>  I've seen images being re-requested within the day

This can particularly happen when your users hit different CDN pops, as they tend to maintain completely independant caches (depends on the CDN variety)

**Reply by Piedone** [reply]
Thanks for the hint, that's a curious case. However, we're not running distributed. BTW the cached images and metadata are both on the shared file system of the Azure Web App, i.e. it's not a local drive but an Azure File Share (this causes the performance bottleneck).

I do contemplate `AzureBlobStorageImageCache`. Why I haven't tried it out yet is because the direct static file access, without ImageSharp, shows similar performance issues under Orchard Core, which seems to be a related problem. I'm hoping to fix both of these in a similar way (like with a custom in-memory caching layer like above, or Output Caching).

Yeah, good point about the CDN, that might be a factor.

**Reply by deanmarcussen** [reply]
cool, the thing to keep in mind when exploring this (I see from the Orchard Core issue you have done a lot of testing), is to be aware that the image metadata for ImageSharp.Web can be read a lot during HEAD requests from CDN's.

I think from memory it has around 5minute volatile caching on it, but if you're outside the window then you're back to system io.

**Reply by Piedone** [reply]
Thanks for the pointer. I was actually wondering why we have a lot of IO still happening for supposedly cached files.

Now that I think about it, actually I should try `AzureBlobStorageImageCache` too, since that would make the image cache persistent, as opposed to starting anew with every deployment. That should alleviate the general IO bottleneck somewhat too.

**Reply by Piedone** [reply]
I'm testing it. In OC this would help BTW: https://github.com/SixLabors/ImageSharp.Web/discussions/351

**Reply by deanmarcussen** [reply]
> Now that I think about it, actually I should try AzureBlobStorageImageCache too, since that would make the image cache persistent, as opposed to starting a new with every deployment. That should alleviate the general IO bottleneck somewhat too.

This suggests you are using containers?

Yes, my experiance is you want to make the cache persistant, or the CDN / browser will constantly be regenerating cache because the timestamps of the cached files will constantly be changing on each deployment.

**Reply by Piedone** [reply]
Not containers but non-container Azure App Services, but we do clean deployments. In the end, the local file cache just lives until the next deployment. But this would also be an issue for containers too, good point, so I added this info to https://github.com/OrchardCMS/OrchardCore/issues/15016, thank you.
